src/runtime_src/hip/api/hip_module.cpp

```diff
@@ -1,64 +1,81 @@
-                      uint32_t gridDimZ, uint32_t blockDimX, uint32_t blockDimY,
-                      uint32_t blockDimZ, uint32_t sharedMemBytes, hipStream_t hStream,
-                      void** kernelParams, void** extra)
+                         uint32_t gridDimZ, uint32_t blockDimX, uint32_t blockDimY,
+                         uint32_t blockDimZ, uint32_t sharedMemBytes, hipStream_t hStream,
+                         void** kernelParams, void** extra)
 {
-  if (!f)
-    throw xrt_core::system_error(hipErrorInvalidResourceHandle, "function is nullptr");
+  throw_invalid_resource_if(!f, "function is nullptr");
+
+  auto func_hdl = reinterpret_cast<function_handle>(f);
+  auto hip_mod = module_cache.get(static_cast<function*>(func_hdl)->get_module());
+  throw_invalid_resource_if(!hip_mod, "module associated with function is unloaded");
+
+  auto hip_func = hip_mod->get_function(func_hdl);
+  throw_invalid_resource_if(!hip_func, "invalid function passed");
 
   throw std::runtime_error("Not implemented");
 }
 
-static hipFunction_t
+static function_handle
 hip_module_get_function(hipModule_t hmod, const char* name)
 {
-  if (!name || strlen(name) == 0)
-    throw xrt_core::system_error(hipErrorInvalidValue, "name is invalid");
+  throw_invalid_value_if((!name || strlen(name) == 0), "name is invalid");
 
-  if (!hmod)
-    throw xrt_core::system_error(hipErrorInvalidResourceHandle, "module is nullptr");
+  throw_invalid_resource_if(!hmod, "module is nullptr");
 
-  throw std::runtime_error("Not implemented");
+  auto mod_hdl = reinterpret_cast<module_handle>(hmod);
+  auto hip_mod = module_cache.get(mod_hdl);
+  throw_invalid_resource_if(!hip_mod, "module not available");
+
+  auto hip_func = std::make_shared<function>(mod_hdl, std::string(name));
+  auto func_hdl = hip_func.get();
+  hip_mod->add_function(func_hdl, std::move(hip_func));
+  return func_hdl;
 }
 
-static void
-hip_module_load_data_ex(hipModule_t* module, const void* image, unsigned int numOptions,
-                    hipJitOption* options, void** optionsValues)
+static module_handle
+create_module(const void* image)
 {
-  if (!module)
-    throw xrt_core::system_error(hipErrorInvalidResourceHandle, "module is nullptr");
-
-  throw std::runtime_error("Not implemented");
+  auto ctx = get_current_context();
+  auto hip_module = std::make_shared<module>(ctx, const_cast<void*>(image));
+  auto module_hdl = hip_module.get();
+  module_cache.add(module_hdl, std::move(hip_module));
+  return module_hdl;
 }
 
-static void
-hip_module_load_data(hipModule_t* module, const void* image)
+static module_handle
+hip_module_load_data_ex(const void* image, unsigned int numOptions,
+                        hipJitOption* options, void** optionsValues)
 {
-  if (!module)
-    throw xrt_core::system_error(hipErrorInvalidResourceHandle, "module is nullptr");
+  // Jit options are ignored for now
+  (void)numOptions;
+  (void)options;
+  (void)optionsValues;
```

Comment:
I have never liked this syntax.  Newer C++ has std::ignore which you can assign to, but personally I always either comment out the named parameters in the function signature or remove it e.g. `void foo(Type /*type*/)` or `void foo(Type)`.  You don't have to make this change, it's up to you.

Reply:
Yes, just removing the parameter names is a good way to go with C++ in this case.